### DISCUSSION TITLE
(treesitter) set highlight for nodes of multiple groups

### DISCUSSION THREAD

**Discussion by ExpandingMan** [neovim/neovim#27538] [Q&A]
For some syntax, when doing `:TSHighlightCapturesUnderCursor` I see multiple groups listed, for example
![image](https://github.com/neovim/neovim/assets/11379353/4f124d13-1ad7-4ff8-bc33-ea20f2baad17)

Is there a way to set highlighting only for syntax which is a member of *both* of these groups?  In this example, `@constant` seems to override whatever was set for `@markup.math`, but what I'm trying to do is set highlighting for `@constant` within `@markup.math` differently.

**Comment by clason** [comment]
First, this is an obsolete command from an archived plugin; you should use the builtin `:Inspect` command.

Second, no. You need to address this on the level of queries, where you need to add a pattern capturing -- with higher precedence! -- the constant inside the math environment. (`:InspectTree` and `:EditQuery` will help you.)

**Comment by ExpandingMan** [comment]
Thanks, think that helps me understand what's going on a bit better.  By any chance are you aware of any examples that use queries in syntax highlighting?  I believe I'd have to write a lua script that uses functions found in `vim.treesitter.query` and then conditionally call `nvim_set_hl` (I probably won't do this because it sounds computationally expensive and none of this is worth the complexity to me, but I'm curious if this is a common use case).

**Reply by clason** [reply]
I think you're missing a lot of basic knowledge on this, which you should first acquire. (None of what you wrote is even remotely correct ;)) It's not trivial, but `:h treesitter-query` and the following section(s) and https://tree-sitter.github.io/tree-sitter/using-parsers#pattern-matching-with-queries should be a start. Ask in https://matrix.to/#/#nvim-treesitter:matrix.org if you need more help.
